activerecord/lib/active_record/persistence.rb

```diff
@@ -782,7 +782,11 @@ def delete
     def destroy
       _raise_readonly_record_error if readonly?
       destroy_associations
-      @_trigger_destroy_callback ||= persisted? && destroy_row > 0
+      if @_trigger_destroy_callback
+        destroy_row
+      else
+        @_trigger_destroy_callback = persisted? && destroy_row > 0
+      end
```

Comment:
Nit:
What I liked with the old version was that it made it [super clear that we should destroy if persisted.](https://github.com/rails/rails/pull/46197/files#diff-11b42664eb9834972953ecd5725c75fd6020af6d4ee5da37ef6d8e0c146f7773L679-L683)

The fact that we should destroy when `@_trigger_destroy_callback` is true incidental. 

Here's an alternative implementation re-establishing the direct relation between persisted and destroy_row.

```suggestion
      destroyed_count = persisted? ? destroy_row : 0
      @_trigger_destroy_callback ||= persisted? && destroyed_count > 0
```